javalib/src/main/scala/java/util/PriorityQueue.scala

```diff
@@ -47,47 +62,74 @@ class PriorityQueue[E] private (
         NaturalComparator.select(c.comparator().asInstanceOf[Comparator[_ >: E]])
       case _ =>
         NaturalComparator
-    }, internal = true)
+    }, internal = true, roundUpToPowerOfTwo(c.size()))
     addAll(c)
   }
 
   def this(c: PriorityQueue[_ <: E]) = {
-    this(c.comp.asInstanceOf[Comparator[_ >: E]], internal = true)
+    this(c.comp.asInstanceOf[Comparator[_ >: E]], internal = true, roundUpToPowerOfTwo(c.size()))
     addAll(c)
   }
 
   def this(sortedSet: SortedSet[_ <: E]) = {
     this(NaturalComparator.select(
         sortedSet.comparator().asInstanceOf[Comparator[_ >: E]]),
-        internal = true)
+        internal = true,
+        roundUpToPowerOfTwo(sortedSet.size()))
     addAll(sortedSet)
   }
 
+  /* Get the best available implementation of inner array for the given platform.
+   *
+   * Use Array[AnyRef] in WebAssembly to avoid JS-interop. In JS, use js.Array.
+   * It is resizable by nature, so manual resizing is not needed.
+   *
+   * `linkTimeIf` is needed here to ensure the optimizer knows
+   * there is only one implementation of `InnerArrayImpl`, and de-virtualize/inline
+   * the function calls.
+   */
+
+  private val innerImpl: InnerArrayImpl =
+    LinkingInfo.linkTimeIf[InnerArrayImpl](LinkingInfo.isWebAssembly) {
+      InnerArrayImpl.JArrayImpl
+    } {
+      InnerArrayImpl.JSArrayImpl
+    }
+
+  private var inner: innerImpl.Repr = innerImpl.make(initialCapacity)
+
+  // Wasm only
   // The index 0 is not used; the root is at index 1.
   // This is standard practice in binary heaps, to simplify arithmetics.
```

Comment:
We should keep these two lines as comments on `private var inner`. They apply to JS as well.

Also, that means that the notion of "capacity" is a skewed in this class. The array should have 1 more elements than the capacity, for "capacity" being the external notion of capacity.

Put differently, it is probably easier to think of "capacity" as the *internal* notion of capacity, which is 1 more than the "public capacity". The public capacity only shows up in `def this(initialCapacity)` and `def this(c: Collection)`.